### ISSUE TITLE
False negatives in MemorySanitizer

### ISSUE THREAD

**Issue by shao-hua-li** [llvm/llvm-project#54613]
MemorySanitizer missed some uninitialized variable reports.

a.c
```c
struct a {
  int b;
  int c
} 
d(struct a *g) {
  struct a f = *g;
  if (f.b)
    printf("f.b\n");
  if (f.c)
    printf("f.c\n");
}
main() {
  struct a e;
  d(&e);
}
```

I tested with clang trunk. You can find that -O1,-O2,-O3 report nothing and have divergent outputs.
```
$ clang -fsanitize=memory -w -O0 a.c; ./a.out
==1533603==WARNING: MemorySanitizer: use-of-uninitialized-value
...
$
$ clang -fsanitize=memory -w -O1 a.c; ./a.out
f.b
$
$ clang -fsanitize=memory -w -O2 a.c; ./a.out
f.b
f.c
$
$ clang -fsanitize=memory -w -O3 a.c; ./a.out
f.b
f.c
$
$ clang -fsanitize=memory -w -Os a.c; ./a.out
==1533603==WARNING: MemorySanitizer: use-of-uninitialized-value
...
$
```
